libpwquality: 1.4.2 -> 1.4.4

06kellyjac opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Updated `libpwquality`

- corrected the license: https://github.com/libpwquality/libpwquality/blob/libpwquality-1.4.4/COPYING
- added long description
- added myself as a maintainer
- removed unnecessary patch (it's part of the releases now) **Edit:** :facepalm: the nixpkgs release is on 1.4.2 not 1.4.4 ...
  - **Edit:** updated
- split outputs
- added pam for building the pam module

This change is done in preparation of adding libpwquality to pam:

![image](https://user-images.githubusercontent.com/9866621/155527880-baf6a1e8-30d9-4e61-9e6b-75effbdd5c32.png)


I just want to check:

- there aren't too many rebuilds and master is fine
- ~that adding pam as a build input doesn't have to be linux only?~
  - broke the mac build but after making pam linux only it builds fine
- that splitting outputs doesn't break any other builds
- that $lib/lib/python3.9/site-packages is the right place for the python stuff
  - without the patch it goes in $out/lib instead separate from all the other lib stuff
  - there also might be a better way to do this than the patch method I've gone with

cc: @baloo since you've modified libpwquality not too long ago
also @jonringer and @jtojnar since you both know how to use outputs properly :slightly_smiling_face:  

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

06kellyjac commented:
Result of `nixpkgs-review pr 161652` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>clevis</li>
    <li>gnome-secrets</li>
    <li>gnome.gnome-control-center</li>
    <li>gnome.gnome-disk-utility</li>
    <li>gnome.gnome-initial-setup</li>
    <li>gnome.seahorse</li>
    <li>libpwquality</li>
    <li>phosh</li>
  </ul>
</details>

06kellyjac commented:
Result of `nixpkgs-review pr 161652` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>clevis</li>
    <li>gnome-secrets</li>
    <li>gnome.gnome-control-center</li>
    <li>gnome.gnome-disk-utility</li>
    <li>gnome.gnome-initial-setup</li>
    <li>gnome.seahorse</li>
    <li>libpwquality</li>
    <li>phosh</li>
  </ul>
</details>

06kellyjac commented:
/marvin opt-in
/status needs_reviewer

jtojnar commented on pkgs/development/libraries/libpwquality/default.nix:
```diff
@@ -1,36 +1,47 @@
+      function and PAM module that can be used instead of pam_cracklib. The
+      module supports all the options of pam_cracklib.
+    '';
+    license = licenses.gpl2Plus;
```
At a glance it looks like it is dual-licensed under `BSD-3-Clause OR GPL-2.0-or-later`. Since Nixpkgs cannot currently express that, I would go with the less restrictive licence of the pair.

06kellyjac replied:
on most of the rust MIT OR Apache 2 we put it in the comments `/* or */` so I'll do it this way

https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/misc/lscolors/default.nix#L20

jtojnar replied:
I do not think that is right. Most systems that check licenses will interpret that as `AND` (e.g. adding an unfree license in the mix would result in the package being considered unfree).

06kellyjac replied:
There's a lot of packages with `/* or */`

```
λ rg "/\* (or|OR) \*/" . | wc -l
106
```

In this case they're both free. If someone is against GPL or BSD3 they can look at the file and make their decision rather than just getting 1 piece of info.

In the case that something is free or unfree (without support for OR) I'd like to at least know something is licensed unfree and then I can review if it's just optional, or I need to avoid certain files and features, or whatever the conditions are and then either ignore or add an overlay. Typically with free and unfree splits we have an overridable argument that can toggle that feature

Until there's native support for conditionals I much prefer it this way

jtojnar commented on pkgs/development/libraries/libpwquality/python-binding-prefix.patch:
```diff
@@ -0,0 +1,13 @@
+ 
+ install-exec-local:
+-	CFLAGS="${CFLAGS} -fno-strict-aliasing" @PYTHONBINARY@ setup.py build --build-base py$(PYTHONREV) install --prefix=${DESTDIR}${prefix}
++	CFLAGS="${CFLAGS} -fno-strict-aliasing" @PYTHONBINARY@ setup.py build --build-base py$(PYTHONREV) install --prefix=${lib}
```
This looks reasonable way to change the output. Though I would keep the `DESTDIR` part, even though we do not use it, to make the diff minimal.

06kellyjac replied:
Updated

jtojnar commented on pkgs/development/libraries/libpwquality/python-binding-prefix.patch:
```diff
@@ -0,0 +1,13 @@
+ 
+ install-exec-local:
+-	CFLAGS="${CFLAGS} -fno-strict-aliasing" @PYTHONBINARY@ setup.py build --build-base py$(PYTHONREV) install --prefix=${DESTDIR}${prefix}
++	CFLAGS="${CFLAGS} -fno-strict-aliasing" @PYTHONBINARY@ setup.py build --build-base py$(PYTHONREV) install --prefix=${lib}
```
Also, it might be better to use a separate `py` output like e.g. libxml2 does.

jtojnar replied:
And something like this:

https://github.com/NixOS/nixpkgs/blob/19d4f7dc485f74109bd66ef74231285ff797a823/pkgs/top-level/python-packages.nix#L4727-L4730

to be able to use `python3.pkgs.libpwquality`

06kellyjac replied:
Added something based on `librepo`

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18714,7 +18714,9 @@ with pkgs;
   libpulsar = callPackage ../development/libraries/libpulsar { };
 
-  libpwquality = callPackage ../development/libraries/libpwquality { };
+  libpwquality = callPackage ../development/libraries/libpwquality {
+    python = python3;
+  };
```
```suggestion
  libpwquality = callPackage ../development/libraries/libpwquality { };
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4665,6 +4665,13 @@ in {
 
+  libpwquality = pipe pkgs.libpwquality [
+    toPythonModule
+    (p: p.overrideAttrs (super: { meta = super.meta // { outputsToInstall = [ "py" ]; }; }))
+    (p: p.override { enablePython = true; inherit python; })
+    (p: p.py)
```
Please add a comment what this is doing.

06kellyjac replied:
It's doing what the others are doing and none of those have comments

SuperSandro2000 replied:
Hmmm... I think it would be easier to understand if we just passed in enablePython and then set meta inside the package with lib.optional.

kevincox replied:
Personally I think this is fine.

SuperSandro2000 commented on pkgs/development/libraries/libpwquality/default.nix:
```diff
@@ -1,36 +1,57 @@
-      url = "https://github.com/libpwquality/libpwquality/pull/40.patch";
-      sha256 = "1ypccq437wxwgddd98cvd330jfm7jscdlzlyxgy05g6yzrr68xyk";
-    })
+  patches = lib.optionals enablePython [
```
```suggestion
  patches = [
```
lets prevent bit rot

06kellyjac committed:
libpwquality: 1.4.2 -> 1.4.4

- corrected the license: https://github.com/libpwquality/libpwquality/blob/libpwquality-1.4.4/COPYING
- added long description
- added myself as a maintainer
- removed unnecessary patch
- split outputs
- added pam for building the pam module

kevincox commented:
Result of `nixpkgs-review pr 161652` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>clevis</li>
    <li>gnome-secrets</li>
    <li>gnome.gnome-control-center</li>
    <li>gnome.gnome-disk-utility</li>
    <li>gnome.gnome-initial-setup</li>
    <li>gnome.seahorse</li>
    <li>libpwquality</li>
    <li>phosh</li>
    <li>python310Packages.libpwquality</li>
    <li>python39Packages.libpwquality</li>
  </ul>
</details>
